### ISSUE TITLE
 Out-of-distribution Detection with Boundary Aware Learning

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#20]
### Paper Link
https://link.springer.com/chapter/10.1007/978-3-031-20053-3_14

## Motivation
<img width="1593" alt="image" src="https://user-images.githubusercontent.com/50613939/210528329-5fb21d44-2603-4055-b935-833df4aae351.png">

- ResNet18 trained on MNIST. Number of neurons of its penultimate layer is set to 2 for feature visualization. The background color represents confidence score given by the ResNet18. **The region far from the blue points gets high confidence score**.
- Classification on two gaussian distribution with a MLP. **Classification net gives OOD regions high confidence which is abnormal.**
- The proposed BAL. **ID regions much higher confidence than OOD regions.**

<img width="1300" alt="image" src="https://user-images.githubusercontent.com/50613939/210529329-c11e4173-ae24-486a-b3a4-18b4d0f3c7dd.png">

- In typical classifiers, the ID and OOD data are confused, and both of them get very high confidence scores.
- With the proposed BAL, the OOD data is assigned with much lower confidence,
allowing more effective OOD detection.

## Methodology
<img width="1422" alt="image" src="https://user-images.githubusercontent.com/50613939/210529835-6b9bd9e2-e566-4aa1-a38c-0e06c18dc9b3.png">

- **Trivial OOD feature**: randomly sampled feature from uniform distribution. 
- Generator $G$ is trained for generating ID data and **synthetic ID feature (as negative)** $f$ can be
got by $G(z, c)$.
- Discriminator $D$ is used for separating ID and OOD features.
- Using FGSM to push the feature $f$ towards the boundary of ID manifold, which gets a much lower score from discriminator.

#### Shuffle Loss 

$$
L_s=\mathbb{E}_{P_f}(\log D(f ; T(c))-\log D(f ; c))
$$

With a shuffle function $T(\cdot)$, the positive pair $\langle f, c\rangle$ is transformed to a negative pair $\langle f, \tilde{c}\rangle$ where $\tilde{c}=T(c) \neq c$ is a mismatched label with feature $f$. 

#### Uniform Loss 

$$
L_u=\lambda_c \cdot \mathbb{E}_{P_{\mathbb{U}}} \log D(\hat{f} ; \hat{c})-\mathbb{E}_{P_f} \log D(f ; c)
$$

#### Discriminator Loss
 
$$
\begin{aligned}
& L_t=-\mathbb{E}_{P_f} \log D(f ; c)-\mathbb{E}_{P_z} \log (1-D(G(z) ; c)) \\
& L_d=L_t+L_s+L_u
\end{aligned}
$$

### FGSM (ICLR2015)
Explaining & Harnessing Adversarial Examples
Chinese help: https://blog.csdn.net/WangHY_XCJ/article/details/109401368
paper: https://arxiv.org/abs/1412.6572

### ODIN (ICLR2018)
ENHANCING THE RELIABILITY OF OUT-OF-DISTRIBUTION IMAGE DETECTION IN NEURAL NETWORKS
https://arxiv.org/abs/1706.02690
https://zhuanlan.zhihu.com/p/102870562

#### 
Temperature Scaling

$$
S_i(\boldsymbol{x} ; T)=\frac{\exp \left(f_i(\boldsymbol{x}) / T\right)}{\sum_{j=1}^N \exp \left(f_j(\boldsymbol{x}) / T\right)}
$$

Using temperature scaling can separate the softmax scores between in- and out-of-distribution images, making out-of-distribution detection effective.

### Input Preprocessing

$$
\tilde{\boldsymbol{x}}=\boldsymbol{x}-\varepsilon \operatorname{sign}\left(-\nabla_{\boldsymbol{x}} \log S_{\hat{y}}(\boldsymbol{x} ; T)\right)
$$

### Out-of-distribution Detector

$$
g(\boldsymbol{x} ; \delta, T, \varepsilon)= \begin{cases}1 & \text { if } \max _i p(\tilde{\boldsymbol{x}} ; T) \leq \delta,  \\ 0 & \text { if } \max _i p(\tilde{\boldsymbol{x}} ; T)>\delta\end{cases}
$$

The parameters $T, \varepsilon$ and $\delta$ are chosen so that the true positive rate (i.e., the fraction of in-distribution images correctly classified as in-distribution images) is $95\%$.

### Generalized ODIN (CVPR2020)

Generalized ODIN: Detecting Out-of-distribution Image without Learning from
Out-of-distribution Data

https://openaccess.thecvf.com/content_CVPR_2020/papers/Hsu_Generalized_ODIN_Detecting_Out-of-Distribution_Image_Without_Learning_From_Out-of-Distribution_Data_CVPR_2020_paper.pdf

Chinese help
https://zhuanlan.zhihu.com/p/110621616

## Experiments 

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/50613939/210681505-76528902-214a-43f7-ae7a-399d0e1e139d.png">

<img width="1175" alt="image" src="https://user-images.githubusercontent.com/50613939/210681549-4f10f342-e4e8-4f34-92ec-37145c6f277c.png">

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/50613939/210681597-acd13d76-65eb-4e3c-8fc1-e13c795b0e38.png">

<img width="1163" alt="image" src="https://user-images.githubusercontent.com/50613939/210681657-f56ec1b5-902b-4562-bf81-dc98168c8a4e.png">

### Ablation 
<img width="1172" alt="image" src="https://user-images.githubusercontent.com/50613939/210681073-8a3d04e5-b82a-4368-aa54-307f16de26ac.png">
